python_modules/dagster/dagster/_core/definitions/asset_condition.py

```diff
@@ -50,7 +63,43 @@ def frozen_metadata(self) -> FrozenSet[Tuple[str, MetadataValue]]:
         return frozenset(self.metadata.items())
 
 
-@whitelist_for_serdes
+def get_serializable_candidate_subset(
+    candidate_subset: Optional[AssetSubset],
+) -> Optional[AssetSubset]:
+    """Do not serialize the candidate subset if it is an AllPartitionsSubset."""
+    if candidate_subset and isinstance(candidate_subset.value, AllPartitionsSubset):
+        return None
+    return candidate_subset
+
+
+class CandidateSubsetSerializer(FieldSerializer):
+    def pack(
+        self,
+        candidate_subset: Optional[AssetSubset],
+        whitelist_map: WhitelistMap,
+        descent_path: str,
+    ) -> Optional[Mapping[str, Any]]:
+        # On all ticks, the root condition starts with an AllPartitionsSubset as the candidate
+        # subset. This would be wasteful to calculate and serialize in its entirety, so we instead
+        # store this as `None` and reconstruct it as needed.
+        # This does mean that if new partitions are added between serialization time and read time,
```

Comment:
This seems like it could cause problems in some edge cases?

Reply:
Imagine on tick N, partition `foo` does not exist for a partitions definition. This means that there is no way for it to have been a part of the `true_subset` of any evaluations (as it did not exist at that time).

On tick N+1, partition `foo` now exists. Assume in the worst case that the candidate subset was `None` for all of the evaluations in the tree. `foo` will now be considered to have been part of the `candidate_subset` of the previous tick for all of these evaluations. However, it will not be in the `true_subset` for any of them. This means that it will be considered to have evaluated to `False` for all evaluations.

This seems like a fairly logical outcome -- any condition about this partition would have been "false", as the partition didn't exist.

Definitely not an ideal situation, but it's not really possible to store the full set of partitions on every tick, at least for things like big DynamicPartitionsDefinitions, and I don't think it's all that bad of an outcome

Reply:
> This seems like a fairly logical outcome -- any condition about this partition would have been "false", as the partition didn't exist.

What about a condition like "is missing"? "false" would imply that it was present (even though it didn't exist).

> Definitely not an ideal situation, but it's not really possible to store the full set of partitions on every tick, at least for things like big DynamicPartitionsDefinitions, and I don't think it's all that bad of an outcome

Yeah, I can brainstorm fancy solutions that might make this less of a burden, but agree that we might need to make this compromise now.

If we are making this compromise, I think it might be preferable to have something like `candidate_subset: Union[PartitionsSubset, SentinelValueThatMeansEverythingForever]`? This would:
- Force any code that uses this attribute to deal with its quirk.
- Avoid the strangeness of serialized values coming back different than when they were serialized.

Reply:
Yeah I don't mind the sentinel value solution -- leaves open the option of showing things like "unknown" in the UI for things like '#false partitions'

Reply:
Added the sentinel value